File: VB/dxSample/Custom Layout Control/CustomLayoutControlItemSkinPainter.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports DevExpress.Skins
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraLayout
Imports DevExpress.XtraLayout.Painting

Namespace dxSample
    Friend Class CustomLayoutControlItemSkinPainter
        Inherits LayoutControlItemSkinPainter

        Protected Overrides Sub PatchImageIndexCore(ByVal e As DevExpress.XtraLayout.ViewInfo.BaseLayoutItemViewInfo, ByVal info As SkinElementInfo, ByVal controlItem As LayoutControlItem)
            info.Bounds = GetPaintBounds(e)
            info.ImageIndex = -1
            If e.OwnerILayoutControl.OptionsView.DrawItemBorders Then
                info.ImageIndex = 0
            End If
            If e.State = ObjectState.Hot AndAlso (e.OwnerILayoutControl.OptionsView.AllowHotTrack) Then
                info.ImageIndex = 1
            End If
            If controlItem IsNot Nothing AndAlso controlItem.Control IsNot Nothing Then
                Dim highlightFocusedItemEx As Boolean = (TryCast(controlItem, CustomLayoutControlItem)).HighlightFocusedItemEx
                If (controlItem.Control.ContainsFocus OrElse IsPopupOpen(controlItem.Control)) AndAlso highlightFocusedItemEx Then
                    info.ImageIndex = 2
                End If
                If (Not controlItem.Control.Enabled) AndAlso (highlightFocusedItemEx) Then
                    info.ImageIndex = 3
                End If
            Else
                If e.OwnerILayoutControl.EnableCustomizationMode Then
                    info.ImageIndex = 3
                End If
            End If
            If info.ImageIndex > 0 AndAlso Not(e.OwnerILayoutControl.OptionsView.DrawItemBorders) Then
                Dim rect As Rectangle = info.Bounds
                rect.Inflate(-1, -1)
                info.Bounds = rect
            End If
            PatchImageIndex(info, e)
            If info.ImageIndex >= 0 Then
                ObjectPainter.DrawObject(e.Cache, SkinElementPainter.Default, info)
            End If
        End Sub
    End Class
End Namespace

File: CS/dxSample/Custom Layout Control/CustomLayoutControlItemSkinPainter.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using DevExpress.Skins;
using DevExpress.Utils.Drawing;
using DevExpress.XtraLayout;
using DevExpress.XtraLayout.Painting;

namespace dxSample {
    internal class CustomLayoutControlItemSkinPainter : LayoutControlItemSkinPainter {
        protected override void PatchImageIndexCore(DevExpress.XtraLayout.ViewInfo.BaseLayoutItemViewInfo e, SkinElementInfo info, LayoutControlItem controlItem) {
            info.Bounds = GetPaintBounds(e);
            info.ImageIndex = -1;
            if (e.OwnerILayoutControl.OptionsView.DrawItemBorders) {
                info.ImageIndex = 0;
            }
            if (e.State == ObjectState.Hot && (e.OwnerILayoutControl.OptionsView.AllowHotTrack)) {
                info.ImageIndex = 1;
            }
            if (controlItem != null && controlItem.Control != null) {
                bool highlightFocusedItemEx = (controlItem as CustomLayoutControlItem).HighlightFocusedItemEx;
                if ((controlItem.Control.ContainsFocus || IsPopupOpen(controlItem.Control)) && highlightFocusedItemEx) {
                    info.ImageIndex = 2;
                }
                if (!controlItem.Control.Enabled && (highlightFocusedItemEx)) {
                    info.ImageIndex = 3;
                }
            } else {
                if (e.OwnerILayoutControl.EnableCustomizationMode) {
                    info.ImageIndex = 3;
                }
            }
            if (info.ImageIndex > 0 && !(e.OwnerILayoutControl.OptionsView.DrawItemBorders)) {
                Rectangle rect = info.Bounds;
                rect.Inflate(-1, -1);
                info.Bounds = rect;
            }
            PatchImageIndex(info, e);
            if (info.ImageIndex >= 0) {
                ObjectPainter.DrawObject(e.Cache, SkinElementPainter.Default, info);
            }
        }
    }
}
